### ISSUE TITLE
git clone fails with transfer.fsckobjects = true: zeroPaddedFilemode

### ISSUE THREAD

**Issue by tzz** [robbyrussell/oh-my-zsh#4963]
Running Git 2.7.4:

```console
% git clone https://github.com/robbyrussell/oh-my-zsh.git .oh-my-zsh 
Cloning into '.oh-my-zsh'...
remote: Counting objects: 15624, done.
error: object 2b7227859263b6aabcc28355b0b994995b7148b6: zeroPaddedFilemode: contains zero-padded file modes
fatal: Error in object
fatal: index-pack failed
```

Workaround: use the tarball/zip archive...

**Comment by mcornella** [comment]
This is what fsck says:
```
$ git fsck
Checking object directories: 100% (256/256), done.
warning in tree 2b7227859263b6aabcc28355b0b994995b7148b6: zeroPaddedFilemode: contains zero-padded file modes
warning in tree a18c4d13c2a5fa2d4ecd5346c50e119b999b807d: zeroPaddedFilemode: contains zero-padded file modes
warning in tree 84df066176c8da3fd59b13731a86d90f4f1e5c9d: zeroPaddedFilemode: contains zero-padded file modes
```

Fixing this might imply rewriting history, but it seems that it's not such a big deal.
Can you check that you have [these fsck settings enabled](https://github.com/holman/dotfiles/issues/231#issuecomment-196412751), and if so, trying again disabling them?

**Comment by tzz** [comment]
It works without fsck, but see https://groups.google.com/forum/m/#!topic/binary-transparency/f-BI4o8HZW0 for the rationale to have that always enabled. FWIW, this is the relevant part of my .gitconfig and yes, it works without it.

```
# for $deity's sake, check that anything we're getting is complete and sane on a regular basis 
[transfer] 
        fsckobjects = true 
[fetch] 
        fsckobjects = true 
[receive] 
        fsckObjects = true
```

**Comment by mcornella** [comment]
From what I gather, just `transfer.fsckobjects` should be enough. I'm trying to find the specific point in time when these where introduced, but I can't assure you that we will fix them because that would be a major breakage for everybody.

**Comment by mcornella** [comment]
You can also ignore `zeroPaddedFilemode` errors by setting `fsck.zeroPaddedFilemode = ignore`, which won't affect your other configurations.

**Comment by tzz** [comment]
I don't know the internals but this may hit other users, see https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=813157 for the discussion where Debian is considering making this the default. So unfortunately this could become a wider problem.

The `fsck.zeroPaddedFilemode = ignore` setting didn't work for me with `transfer.fsckobjects` set to true, sorry. Git 2.7.4. `man git` says that the general `fsck.msg-id` setting is supported for this kind of use case but I can't find any online cites of the specific solution you suggested so maybe the message ID is wrong? Or 2.7.4 is broken in some way? How did you verify your solution, and with what Git version?

```
% git clone -c fsck.zeroPaddedFilemode=ignore https://github.com/robbyrussell/oh-my-zsh.git /tmp/o
Cloning into '/tmp/o'...
remote: Counting objects: 15624, done.
error: object 2b7227859263b6aabcc28355b0b994995b7148b6: zeroPaddedFilemode: contains zero-padded file modes
fatal: Error in object
fatal: index-pack failed
```

**Comment by mcornella** [comment]
> The fsck.zeroPaddedFilemode = ignore setting didn't work for me with transfer.fsckobjects set to true, sorry. Git 2.7.4.

Sorry, I should've been more explicit. That's  another `git config` setting, see https://github.com/collectiveidea/delayed_job/issues/897#issuecomment-200891999.

> I don't know the internals but this may hit other users, see https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=813157 for the discussion where Debian is considering making this the default. So unfortunately this could become a wider problem.

I'll review that, thank you for the heads up!

**Comment by mcornella** [comment]
Ok, I've finished reviewing what should we change and it doesn't look good. 

I've used the [`fast-export + fast-import` trick](http://stackoverflow.com/a/14839059/5798232) to fix the fsck errors, but a lot of SHA-1 checksums have been modified. The most recent commit that hasn't changed is 9fde1531f379449faa4f103d8493e8d2665e3055, from 2009. So there's no way we are changing all these commits (3750!) for such a trivial problem. We will have to instruct people to add the ignore zeroPaddedFilemode setting.

Below is a difference of both repository's histories:

The commands I used to compare them:
```zsh
pushd oh-my-zsh_old; git --oneline --no-abbrev-commit master > ../omz_history_old.log; popd
pushd oh-my-zsh_new; git --oneline --no-abbrev-commit master > ../omz_history_new.log; popd
diff omz_history* > omz_history.diff
```

[and the diff between the two repositories](https://gist.github.com/mcornella/4a9bd27b96af3fd1ebc5c2e6ff678ed2) (GitHub won't let me post the full diff here, it's bigger than 650kB).

**Comment by mcornella** [comment]
**TL;DR** the solution to this issue is adding the following:
```
[fsck]
    zeroPaddedFilemode = ignore
```
to your `~/.gitconfig` file.

**Comment by mcornella** [comment]
That setting doesn't seem to work for me. Investigating...

**Comment by mcornella** [comment]
I've found the official documentation for that `fsck.zeroPaddedFilemode` setting, it's at [git-config](https://git-scm.com/docs/git-config/2.6.0) (look for `fsck.<msg-id>`). 

Now, I've seen this setting works when doing a `git fsck` manually: the error message doesn't appear again. But it still happens with the `fsckobjects = true` setting, so I have the suspicion that there's a bug here.

Investigating further...

**Comment by mcornella** [comment]
Ok, so it seems there's a bug on git's behavior when fetching: https://github.com/holman/dotfiles/issues/231#issuecomment-210259671.

I can't do anything else but recommend you to override those `*.fsckobjects` settings (set them to false), in just the OMZ repository at least.

There's no other solution to that annoying error that doesn't involve everybody's OMZ setups catching fire.

**Comment by tzz** [comment]
Thanks for looking into it. I understand completely. When that Git deficiency is fixed in http://thread.gmane.org/gmane.comp.version-control.git/288336 I will set things up as you recommend, though by then the zero padded file mode will probably not be an error anymore due to user complaints :)

**Comment by kevinburke** [comment]
FWIW, if you want to work around this and leave your global git config alone, you can follow these instructions: http://stackoverflow.com/q/41029654/329700

**Comment by CristianCantoro** [comment]
@johnramsden:
> It becomes a huge problem when attempting to use automation tools such as package managers or built systems.

I have been using this in my scripts (see for example [install_ohmyzsh.sh#L70](https://github.com/CristianCantoro/cms_deploy/blob/master/provision/zsh/install_ohmyzsh.sh#L70)):
```bash
git clone --depth=1 https://github.com/robbyrussell/oh-my-zsh.git
```
It does not clone the whole repo history, just the latest version, which should be ok anyway for automated installs.
(I am quite sure I did not come up with this myself, but I am unable to retrace it to the original source).

Alternatively there is the command suggested in this [StackOverflow answer](https://stackoverflow.com/questions/41029654/ignore-fsck-zero-padded-file-mode-errors-in-git-clone), pointed out above by @kevinburke:
```
git clone --config transfer.fsckobjects=false \
    --config receive.fsckobjects=false \
    --config fetch.fsckobjects=false \
    git://github.com/robbyrussell/oh-my-zsh.git
```

**Comment by johnramsden** [comment]
@CristianCantoro That's what I've been doing in my own scripts, disabling fsck on omz only. The real problem is when you attempt to use package managers you have to manually go and change the pull commands every time which is really inconvenient. For this reason alone I am just pulling omz directly instead of using the package manager.

It's just too bad there isn't a better option to fix the problem. A fix really needs to be integrated into git it says a corrupted repo has been fixed and lets you use the fixed version transparently. I'll admit I don't know much about the internals of git so I don't even know if that's possible.
